streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -488,7 +491,12 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
           registerReceiver(streamId, typ, host, executorId, receiverEndpoint, context.senderAddress)
         context.reply(successful)
       case AddBlock(receivedBlockInfo) =>
-        context.reply(addBlock(receivedBlockInfo))
+        if (WriteAheadLogUtils.isBatchingEnabled(ssc.conf, isDriver = true)) {
+          val f = Future(addBlock(receivedBlockInfo))(walBatchingThreadPool)
+          f.onComplete(result => context.reply(result.get))(walBatchingThreadPool)
```

Comment:
Right, if it is running. So from what I can understand the code in the `Future` itself runs in one thread, and the `onComplete` code runs in another. Are both these no longer safe to run or is it safe to add the block - my take is both are not safe to run.

We should probably need synchronization to ensure this safety then